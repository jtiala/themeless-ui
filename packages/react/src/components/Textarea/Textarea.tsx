import { CommonComponentProps, cn } from "@themeless-ui/utils";
import { TextareaHTMLAttributes } from "react";

type HTMLTextareaProps = TextareaHTMLAttributes<HTMLTextAreaElement>;

export type TextareaProps = {
  /**
   * Name of the textarea.
   */
  name?: HTMLTextareaProps["name"];

  /**
   * Brief hint for the user as to what kind of information is expected in the textarea.
   */
  placeholder?: HTMLTextareaProps["placeholder"];

  /**
   * The value of the textarea.
   * Use only when using the textarea as controlled component (state handled by React).
   */
  value?: HTMLTextareaProps["value"];

  /**
   * Prevent the user from interacting with the textarea.
   */
  disabled?: HTMLTextareaProps["disabled"];

  /**
   * Indicates that the user should not be able to edit the value of the textarea.
   */
  readOnly?: HTMLTextareaProps["readOnly"];

  /**
   * Indicates that the user must specify a value for the textarea before the owning form can be submitted.
   */
  required?: HTMLTextareaProps["required"];

  /**
   * Callback to call when the textarea's value is modified.
   */
  onChange?: HTMLTextareaProps["onChange"];

  /**
   * Callback to call when the textarea is clicked.
   */
  onClick?: HTMLTextareaProps["onClick"];

  /**
   * Callback to call when the textarea receives focus.
   */
  onFocus?: HTMLTextareaProps["onFocus"];

  /**
   * Callback to call when the textarea loses focus.
   */
  onBlur?: HTMLTextareaProps["onBlur"];
} & CommonComponentProps;

const className = cn("textarea");

/**
 * Display an textarea.
 */
export function Textarea({
  name,
  placeholder,
  value,
  disabled,
  readOnly,
  required,
  onChange,
  onClick,
  onFocus,
  onBlur,
  id,
  testId,
}: TextareaProps) {
  return (
    <textarea
      name={name}
      placeholder={placeholder}
      value={value}
      disabled={disabled}
      readOnly={readOnly}
      required={required}
      onChange={onChange}
      onClick={onClick}
      onFocus={onFocus}
      onBlur={onBlur}
      className={className}
      id={id}
      data-testid={testId}
    />
  );
}
