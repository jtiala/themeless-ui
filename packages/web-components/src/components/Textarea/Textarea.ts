import style from "@themeless-ui/style/dist/textarea.css?inline";
import { cn } from "@themeless-ui/utils";
import { html, nothing, unsafeCSS } from "lit";
import { customElement, property } from "lit/decorators.js";
import {
  TUIBlurEvent,
  TUIChangeEvent,
  TUIClickEvent,
  TUIFocusEvent,
  TUIMouseEnterEvent,
  TUIMouseLeaveEvent,
  TUITouchEndEvent,
  TUITouchStartEvent,
} from "../../events";
import { TUIComponent } from "../TUIComponent";

const className = cn("textarea");

@customElement("tui-textarea")
export class Textarea extends TUIComponent {
  /**
   * Name of the textarea.
   */
  @property()
  name?: HTMLTextAreaElement["name"];

  /**
   * Prevent the user from interacting with the textarea.
   */
  @property({ type: Boolean })
  disabled?: HTMLTextAreaElement["disabled"];

  /**
   * Indicates that the user should not be able to edit the value of the textarea.
   */
  @property({ type: Boolean })
  readOnly?: HTMLTextAreaElement["readOnly"];

  /**
   * Indicates that the user must specify a value for the textarea before the owning form can be submitted.
   */
  @property({ type: Boolean })
  required?: HTMLTextAreaElement["required"];

  /**
   * Brief hint for the user as to what kind of information is expected in the textarea.
   */
  @property()
  placeholder?: HTMLTextAreaElement["placeholder"];

  /**
   * The value of the textarea.
   * Use only when using the textarea as controlled component (state handled by React).
   */
  @property()
  value?: HTMLTextAreaElement["value"];

  private _handleChange() {
    this.dispatchEvent(new TUIChangeEvent());
  }

  private _handleClick() {
    this.dispatchEvent(new TUIClickEvent());
  }

  private _handleFocus() {
    this.dispatchEvent(new TUIFocusEvent());
  }
  private _handleBlur() {
    this.dispatchEvent(new TUIBlurEvent());
  }

  private _handleMouseEnter() {
    this.dispatchEvent(new TUIMouseEnterEvent());
  }

  private _handleMouseLeave() {
    this.dispatchEvent(new TUIMouseLeaveEvent());
  }

  private _handleTouchStart() {
    this.dispatchEvent(new TUITouchStartEvent());
  }

  private _handleTouchEnd() {
    this.dispatchEvent(new TUITouchEndEvent());
  }

  static styles = unsafeCSS(style);

  override render() {
    return html`
      <textarea
        name="${this.name || nothing}"
        placeholder="${this.placeholder || nothing}"
        value="${this.value || nothing}"
        ?disabled="${this.disabled}"
        ?readOnly="${this.readOnly}"
        ?required="${this.required}"
        @change="${this._handleChange}"
        @click="${this._handleClick}"
        @focus="${this._handleFocus}"
        @blur="${this._handleBlur}"
        @mouseEnter="${this._handleMouseEnter}"
        @mouseLeave="${this._handleMouseLeave}"
        @touchStart="${this._handleTouchStart}"
        @touchEnd="${this._handleTouchEnd}"
        class="${className}"
        id="${this.id || nothing}"
        data-testid="${this.testId || nothing}"
      ></textarea>
    `;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    "tui-textarea": Textarea;
  }
}
