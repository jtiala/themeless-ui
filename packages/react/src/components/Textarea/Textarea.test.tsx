import {
  componentToJson,
  describe,
  expect,
  it,
  render,
  renderer,
  screen,
  userEvent,
  vi,
} from "../../../test";
import { Textarea } from "./Textarea";

const onChange = vi.fn();
const onClick = vi.fn();
const onFocus = vi.fn();
const onBlur = vi.fn();

const defaultTextarea = (
  <Textarea
    onChange={() => onChange()}
    onClick={() => onClick()}
    onFocus={() => onFocus()}
    onBlur={() => onBlur()}
    testId="textarea"
  />
);

const disabledTextarea = (
  <Textarea
    disabled
    onChange={() => onChange()}
    onClick={() => onClick()}
    onFocus={() => onFocus()}
    onBlur={() => onBlur()}
    testId="textarea"
  />
);

const readOnlyTextarea = (
  <Textarea
    readOnly
    onChange={() => onChange()}
    onClick={() => onClick()}
    onFocus={() => onFocus()}
    onBlur={() => onBlur()}
    testId="textarea"
  />
);

describe("Textarea", async () => {
  it("should render the textarea", () => {
    expect(componentToJson(renderer.create(defaultTextarea))).toMatchSnapshot();

    render(defaultTextarea);
    const textarea = screen.getByTestId("textarea");

    expect(textarea).toBeInTheDocument();
  });

  it("should call the callbacks", async () => {
    onChange.mockReset();
    onClick.mockReset();
    onFocus.mockReset();
    onBlur.mockReset();

    render(defaultTextarea);
    const textarea = screen.getByTestId("textarea");

    expect(textarea).toBeInTheDocument();
    expect(onChange.mock.calls.length).toBe(0);
    expect(onClick.mock.calls.length).toBe(0);
    expect(onFocus.mock.calls.length).toBe(0);
    expect(onBlur.mock.calls.length).toBe(0);

    await userEvent.click(textarea);
    expect(onChange.mock.calls.length).toBe(0);
    expect(onClick.mock.calls.length).toBe(1);
    expect(onFocus.mock.calls.length).toBe(1);

    textarea.blur();
    expect(onBlur.mock.calls.length).toBe(1);

    textarea.focus();
    expect(onFocus.mock.calls.length).toBe(2);

    await userEvent.type(textarea, "test{enter}test");
    expect(onChange.mock.calls.length).toBe(9);
    expect(textarea).toHaveValue("test\ntest");
  });

  it("shouldn't be possible to access a disabled textarea", async () => {
    onChange.mockReset();
    onClick.mockReset();
    onFocus.mockReset();
    onBlur.mockReset();

    render(disabledTextarea);
    const textarea = screen.getByTestId("textarea");

    expect(textarea).toBeInTheDocument();
    expect(onChange.mock.calls.length).toBe(0);
    expect(onClick.mock.calls.length).toBe(0);
    expect(onFocus.mock.calls.length).toBe(0);
    expect(onBlur.mock.calls.length).toBe(0);

    await userEvent.click(textarea);
    expect(onChange.mock.calls.length).toBe(0);
    expect(onClick.mock.calls.length).toBe(0);
    expect(onFocus.mock.calls.length).toBe(0);

    textarea.blur();
    expect(onBlur.mock.calls.length).toBe(0);

    textarea.focus();
    expect(onFocus.mock.calls.length).toBe(0);

    await userEvent.type(textarea, "test{enter}test");
    expect(onChange.mock.calls.length).toBe(0);
  });

  it("shouldn't be possible to change the value of a read only textarea", async () => {
    onChange.mockReset();
    onClick.mockReset();
    onFocus.mockReset();
    onBlur.mockReset();

    render(readOnlyTextarea);
    const textarea = screen.getByTestId("textarea");

    expect(textarea).toBeInTheDocument();
    expect(onChange.mock.calls.length).toBe(0);
    expect(onClick.mock.calls.length).toBe(0);
    expect(onFocus.mock.calls.length).toBe(0);
    expect(onBlur.mock.calls.length).toBe(0);

    await userEvent.click(textarea);
    expect(onChange.mock.calls.length).toBe(0);
    expect(onClick.mock.calls.length).toBe(1);
    expect(onFocus.mock.calls.length).toBe(1);

    textarea.blur();
    expect(onBlur.mock.calls.length).toBe(1);

    textarea.focus();
    expect(onFocus.mock.calls.length).toBe(2);

    await userEvent.type(textarea, "test{enter}test");
    expect(onChange.mock.calls.length).toBe(0);
  });
});
