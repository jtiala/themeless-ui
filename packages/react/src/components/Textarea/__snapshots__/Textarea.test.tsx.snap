// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Textarea > should render the textarea 1`] = `
<textarea
  className="tui__textarea"
  data-testid="textarea"
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
/>
`;
